File: java/20/ValidParentheses.java
import java.util.Stack;

/**
 * @author: create by nadao
 * @version: v1.0
 * @description: 有效的括号
 * @date:2019/11/19
 */
public class ValidParentheses {
    public boolean isValid(String s) {
        Stack<Character> mark = new Stack<>();
        for(char cur : s.toCharArray()){
            if(cur == '(' || cur == '[' || cur == '{' ){
                mark.push(cur);
            }
            if(cur == ')' || cur == ']' || cur == '}'){
                if(mark.isEmpty()) {
                    return false;
                }
                char p = mark.pop();
                if (cur == ')' && p != '('){
                    return false;
                }
                if (cur == ']' && p != '['){
                    return false;
                }
                if (cur == '}' && p != '{'){
                    return false;
                }
            }
        }
        return mark.isEmpty();
    }
}

File: python/20/Valid_Parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """

        if len(s) == 0:
            return True
        if len(s) & 1: 
            return False

        stack = []
        m = {'}': '{', ')': '(', ']': '['}

        for c in s:
            if c in m.values():
                stack.append(c)
            elif c in m.keys():
                if len(stack) != 0 and m[c] == stack[-1]:
                    stack.pop()
                else:
                    return False
            else:
                return False
        return len(stack) == 0
